File: solutions/prospective/algorithms/patterns/py/sliding_window_variable.py
"""
Variable Sliding Window Template

Used for finding longest/shortest subarray meeting a condition.
1. Expand right pointer.
2. Shrink left pointer while condition is violated.
"""
=== MEMO START ===
def sliding_window_variable(arr):
    l = 0
    longest = 0
    window_state = set() # or count map
    
    for r in range(len(arr)):
        while arr[r] in window_state:
            window_state.remove(arr[l])
            l += 1
        window_state.add(arr[r])
        longest = max(longest, r - l + 1)
        
    return longest

File: solutions/prospective/algorithms/patterns/js/sliding_window_variable.js
/**
 * Variable Sliding Window Template
 * 
 * Used for finding longest/shortest subarray meeting a condition.
 * 1. Expand right pointer.
 * 2. Shrink left pointer while condition is violated.
 */
// === MEMO START ===
function slidingWindowVariable(arr) {
    let l = 0;
    let longest = 0;
    const windowState = new Set();
    
    for (let r = 0; r < arr.length; r++) {
        while (windowState.has(arr[r])) {
            windowState.delete(arr[l]);
            l++;
        }
        windowState.add(arr[r]);
        longest = Math.max(longest, r - l + 1);
    }
    
    return longest;
}

